### PR TITLE
SBG device 

### PR DESCRIPTION
**Pull request by NoahSprenger**
This PR covers the implementation of the SBG IMU/GNSS.

### CODE REVIEW

**Comment by NoahSprenger** [comment]
Review needs to cover the uses of unsafe.

File: Cargo.toml [hunk 1]
```diff
@@ -24,8 +24,8 @@ features = ["critical-section-single-core"]
 lto = false
 
 # Only optimize dependencies for size in debug, keeping the top crate debug friendly
-[profile.dev.package."*"]
-opt-level = "s"
+# [profile.dev.package."*"]
```
Comment (jonapap) [inline Cargo.toml:27]: Uncomment this

File: boards/main/Cargo.toml [hunk 2]
```diff
@@ -12,10 +12,12 @@ cortex-m-rtic = "1.1.3"
 panic-halt = "0.2.0"
 fugit = "0.3.3"
 systick-monotonic = "1.0.1"
-defmt-rtt = "0.4.0"
-defmt = "0.3.2"
+defmt-rtt = "^0.4.0"
```
Comment (jonapap) [inline boards/main/Cargo.toml:15]: Why

File: libraries/common-arm/src/sd/sd_interface.rs [hunk 3]
```diff
@@ -158,3 +158,5 @@ impl SdInterface {
             .close_dir(&self.volume, self.root_directory);
     }
 }
+
+unsafe impl Send for SdInterface {}
```
Comment (jonapap) [inline libraries/common-arm/src/sd/sd_interface.rs:162]: Safety comment

File: libraries/sbg-rs/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+nb = "1.0.0"
+defmt = "0.3.2"
+atsamd-hal = { workspace = true }
+cortex-m = { workspace = true }
+cortex-m-rt = "^0.7.0"
+messages = { path = "../../libraries/messages" }
+common-arm = { path = "../../libraries/common-arm" }
+
+[build-dependencies]
+#tinyrlibc = "0.3.0"
```
Comment (jonapap) [inline libraries/sbg-rs/Cargo.toml:20]: Also here

File: libraries/sbg-rs/build.rs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        .join("build");
+
+    // panic!("{}", dst.display());
+
+    println!("cargo:rustc-link-search={}", dst.display());
+    println!("cargo:rustc-link-lib=static=sbgECom");
+
+    println!("cargo:rerun-if-changed=wrapper.h");
+
+    // let bindings = bindgen::Builder::default()
```
Comment (jonapap) [inline libraries/sbg-rs/build.rs:19]: Cleanup this file

File: libraries/sbg-rs/src/data.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+ff 5a 06 00 20 00 c8 6f d5 01 33 9d b6 3c 01 93 4a 3c 52 5c 90 bf 31 a6 b1 3b 6c a0 b1 3b d6 48 9b 3c 12 03 00 00 89 2e 33
```
Comment (jonapap) [inline libraries/sbg-rs/src/data.txt:1]: Delete this

File: libraries/sbg-rs/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+#![no_std]
+
+#![allow(dead_code)]
+#![allow(non_snake_case)]
+#![allow(non_camel_case_types)]
+#![allow(non_upper_case_globals)]
+
+mod bindings;
+
+pub mod sbg;
```
Comment (jonapap) [inline libraries/sbg-rs/src/lib.rs:10]: Maybe add documentation here, giving a quick overview of the purpose of this library?

File: libraries/sbg-rs/src/sbg.rs [hunk 8]
```diff
@@ -0,0 +1,491 @@
...
+use core::slice::{from_raw_parts, from_raw_parts_mut};
+use core::sync::atomic::AtomicUsize;
+use defmt::{debug, flush, info, warn};
+use embedded_hal::serial::Write;
+use hal::gpio::{PA08, PA09, PB16, PB17};
+use hal::sercom::uart::Duplex;
+use hal::sercom::uart::{self, EightBit, Uart};
+use hal::sercom::{IoSet1, Sercom0, Sercom5};
+use messages::sensor::Sbg;
+type Pads = uart::PadsFromIds<Sercom0, IoSet1, PA09, PA08>;
```
Comment (jonapap) [inline libraries/sbg-rs/src/sbg.rs:26]: Separate the types from the rest

File: libraries/sbg-rs/src/sbg.rs [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+        mut bytesToRead: usize,
+    ) -> _SbgErrorCode {
+        if pBuffer.is_null() {
+            return _SbgErrorCode_SBG_NULL_POINTER;
+        }
+        if pBytesRead.is_null() {
+            return _SbgErrorCode_SBG_NULL_POINTER;
+        }
+        // Safe because we check if pBuffer is null and the SBGECom library ensures the buffer given is of the correct size
+        unsafe {
```
Comment (jonapap) [inline libraries/sbg-rs/src/sbg.rs:256]: Remove this big block

File: libraries/sbg-rs/src/sbg.rs [hunk 10]
```diff
@@ -0,0 +1,491 @@
...
+            bytesToRead = SBG_BUFFER_SIZE - index;
+            if bytesToRead == 0 {
+                // Safe because we check if pBytesRead is null
+                    *pBytesRead = 0;
+                return _SbgErrorCode_SBG_READ_ERROR; // no data
+            }
+            let end = bytesToRead + index;
+            // Could be unsafe  
+            array[0..bytesToRead - 1].copy_from_slice( &BUF[index..end - 1] );
+            // Could be unsafe
```
Comment (jonapap) [inline libraries/sbg-rs/src/sbg.rs:272]: Could -> IS UNSAFE :fire:

File: libraries/sbg-rs/src/sbg.rs [hunk 11]
```diff
@@ -0,0 +1,491 @@
...
+                *pBytesRead = bytesToRead;
+            return _SbgErrorCode_SBG_NO_ERROR;
+        }
+        let end = bytesToRead + index;
+        // Could be unsafe
+        array[0..bytesToRead - 1].copy_from_slice(&BUF[index..end - 1] );
+        // Could be unsafe
+            *BUF_INDEX.get_mut() = index + bytesToRead;
+        
+        // Safe because we check if pBytesRead is null
```
Comment (jonapap) [inline libraries/sbg-rs/src/sbg.rs:284]: :+1:

File: libraries/sbg-rs/src/sbg.rs [hunk 12]
```diff
@@ -0,0 +1,491 @@
...
+        return;
+    }
+    // Safe since we check that the pointers are not null
+        let file = CStr::from_ptr(pFileName).to_str().unwrap();
+        let function = CStr::from_ptr(pFunctionName).to_str().unwrap();
+        let _category = CStr::from_ptr(pCategory).to_str().unwrap();
+        let _format = CStr::from_ptr(pFormat).to_str().unwrap();
+    
+    match logType {
+        // silently handle errors
```
Comment (jonapap) [inline libraries/sbg-rs/src/sbg.rs:458]: Add an issue for this

File: boards/main/src/main.rs [hunk 13]
```diff
@@ -1,20 +1,25 @@
 #![no_std]
 #![no_main]
+#[link(name = "c", kind = "static")]
+extern "C" {}
```
Comment (jonapap) [inline boards/main/src/main.rs:4]: Check if we still need this

File: boards/main/src/main.rs [hunk 14]
```diff
@@ -23,23 +28,44 @@ use messages::*;
...
 use systick_monotonic::*;
-
-type Pads = uart::PadsFromIds<Sercom1, IoSet1, PA17, PA16>;
+type Pads = uart::PadsFromIds<Sercom3, IoSet1, PA23, PA22>;
+type PadsSBG = uart::PadsFromIds<Sercom0, IoSet1, PA09, PA08>;
 type Config = uart::Config<Pads, EightBit>;
-
-#[rtic::app(device = hal::pac, peripherals = true, dispatchers = [EVSYS_0])]
+type ConfigSBG = uart::Config<PadsSBG, EightBit>;
+use core::sync::atomic::AtomicU8;
```
Comment (jonapap) [inline boards/main/src/main.rs:35]: Separate the types from the use

File: boards/main/src/main.rs [hunk 15]
```diff
@@ -23,23 +28,44 @@ use messages::*;
...
+    BufferPair<Uart<ConfigSBG, uart::RxDuplex>, SBGBuffer>,
+>;
+const SBG_BUFFER_SIZE: usize = 4096;
+type SBGBuffer = &'static mut [u8; SBG_BUFFER_SIZE];
+static mut BUF_DST: SBGBuffer = &mut [0; SBG_BUFFER_SIZE];
+static mut BUF_DST2: SBGBuffer = &mut [0; SBG_BUFFER_SIZE];
+
+#[rtic::app(device = hal::pac, peripherals = true, dispatchers = [EVSYS_0, EVSYS_1, EVSYS_2])]
 mod app {
+    use hal::{dmac::Transfer, sercom::Sercom};
```
Comment (jonapap) [inline boards/main/src/main.rs:50]: Move to top module

File: boards/main/src/main.rs [hunk 16]
```diff
@@ -70,57 +119,196 @@ mod app {
...
         loop {
             rtic::export::wfi();
         }
     }
 
+    /**
+     * Setups the SBG to output the appropriate data.
+     */
+    #[task(priority = 3, shared = [sbg])]
+    fn sbg_init(mut cx: sbg_init::Context) {
```
Comment (jonapap) [inline boards/main/src/main.rs:251]: Can this be moved to init?

File: boards/main/src/main.rs [hunk 17]
```diff
@@ -190,3 +385,36 @@ mod app {
         });
     }
 }
+
+#[no_mangle]
+pub extern "C" fn _sbrk() {}
```
Comment (jonapap) [inline boards/main/src/main.rs:390]: Double check that this is still needed

**Review by jonapap — APPROVED** [review]

File: boards/main/src/main.rs (hunk 17 above)
Comment (NoahSprenger) [reply boards/main/src/main.rs:390]: Still needed.

**Pull request merged**